src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationSource.cs

```diff
@@ -63,7 +63,21 @@ public abstract class FileConfigurationSource : IConfigurationSource
         /// <param name="builder">The <see cref="IConfigurationBuilder"/>.</param>
         public void EnsureDefaults(IConfigurationBuilder builder)
         {
-            if (FileProvider is null && builder.GetUserDefinedFileProvider() is null)
+            // There are two in-box implementations of IConfigurationBuilder: ConfigurationBuilder and ConfigurationManager.
+            // 1) ConfigurationBuilder.Build takes a list of IConfigurationSources and builds IConfigurationProvider from them.
+            // IConfigurationProvider abstraction does not expose a reference to IConfigurationSources, but most of it implementations do.
+            // ConfigurationBuilder.Build creates an instance of ConfigurationRoot by passing it just the list of providers (NOT the sources).
+            // ConfigurationRoot implements IDisposable and it has only a list of providers (no sources).
+            // When it gets disposed, all providers AND sources that were used to create the providers should be disposed.
+            // This is why FileConfigurationProvider disposes the source when it knows that it's not owned by the user.
+            // 2) ConfigurationManager is also IDisposable, but it has references both sources and providers.
+            // When sources change, it creates new providers and disposes the old ones.
+            // It must not dispose the sources! That is why, in such scenario OwnsFileProvider is set to false.
+            if (builder is IConfigurationManager)
```

Comment:
Is this a safe assumption?  Can't any `FileConfigurationSource` have `Build` called more than once?

Also, I've noticed others implement IConfigurationBuilder before.  Some are forks of ConfigurationManager, others are unique.  I can share some data offline.

A couple other options to consider: 
1. Just create a new FileProvider if the previous one is disposed.  The logic for creating a new one isn't specific to the provider (it's implemented in an extension method we own IIRC).
2. Share a single FIleProvider for default and don't dispose it.  We still leak until finalization, but only one is leaked.
3. Implement ref-counting, or disposal pattern for the source.  If we just add IDispose to the source impl, will it get called by whichever component manages it's lifetime today?

Reply:
@ericstj Thank you for feedback! A lot of good questions! Here are my findings:

> Is this a safe assumption?  Can't any `FileConfigurationSource` have `Build` called more than once?

`ConfigurationManager` can call it more than once. It's called by [ReloadSources](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs#L136) which is called... anytime a source is being deleted, inserted or **any** property changes

> Also, I've noticed others implement IConfigurationBuilder before. Some are forks of ConfigurationManager, others are unique.

`ConfigurationManager` is sealed, I could change this condition from to check for explicit type.

https://github.com/dotnet/runtime/blob/a9717630e4e72fd86690762c1b8bf089a5e91f28/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs#L25

Reply:
> Just create a new FileProvider if the previous one is disposed. 

I was thinking about it. The idea I had was to null the `FileProvider` after disposing it and let `ReloadSources` re-assign it:

https://github.com/dotnet/runtime/blob/a9717630e4e72fd86690762c1b8bf089a5e91f28/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs#L146

https://github.com/dotnet/runtime/blob/1a7904ee34fe29d82fc605e0844f00e238d9d41b/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationSource.cs#L18-L20

https://github.com/dotnet/runtime/blob/1a7904ee34fe29d82fc605e0844f00e238d9d41b/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationSource.cs#L71

But then I've realized that `ReloadSources` creates new providers firsts, then tries to **replace** (it needs new ones to get rid of old ones) them:

https://github.com/dotnet/runtime/blob/a9717630e4e72fd86690762c1b8bf089a5e91f28/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs#L136-L157

and it tries to do that in thread-safe way:

https://github.com/dotnet/runtime/blob/a9717630e4e72fd86690762c1b8bf089a5e91f28/src/libraries/Microsoft.Extensions.Configuration/src/ReferenceCountedProvidersManager.cs#L41-L58

Changing the order would most likely be possible, but I don't see a possibility to do it in a simple way with low risk (I was searching for a fix that could be backported to 8.0)

Reply:
> Share a single FIleProvider for default and don't dispose it. We still leak until finalization, but only one is leaked.

This is something I've not considered. How would it affect perf on various OSes? Is it better to have a single instance of `PhysicalFilesWatcher` rather than many?

https://github.com/dotnet/runtime/blob/1a7904ee34fe29d82fc605e0844f00e238d9d41b/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFileProvider.cs#L32

Currently a new instance if being allocated if the user has not configured a default one (but I don't know whether for example ASP.NET does not do that somewhere else):

https://github.com/dotnet/runtime/blob/a9717630e4e72fd86690762c1b8bf089a5e91f28/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationExtensions.cs#L44

Reply:
>>  Is this a safe assumption? Can't any FileConfigurationSource have Build called more than once?
> ...

> ConfigurationManager is sealed, I could change this condition from to check for explicit type.


My point was that special casing `ConfigurationManager` isn't really a fix.  Anyone can call 

```c#
IConfigurationSource source = new DerivedFileConfigurationSource();
using (IConfigurationProvider provider = source.Build(builder)) { }
using (IConfigurationProvider provider = source.Build(builder)) { }
```
I believe it's intentional that Sources can produce multiple providers.  So we really shouldn't assume that the provider can dispose resources held by the source.

>> Just create a new FileProvider if the previous one is disposed.
> ...
> Changing the order would most likely be possible, but I don't see a possibility to do it in a simple way with low risk (I was searching for a fix that could be backported to 8.0)

Maybe you don't need to change the order - if your design is to delegate the lifetime of the FileProvider to the IConfigurationProvider then you should really be creating a new FileProvider for each call to Build.  I think this is the solution that's most consistent with the change you made.

>> Share a single FIleProvider for default and don't dispose it. We still leak until finalization, but only one is leaked.
> This is something I've not considered. How would it affect perf on various OSes? Is it better to have a single instance of PhysicalFilesWatcher rather than many?

I think it's a tradeoff (but @Jozkee might have a better idea as he's our resident FSW expert).  I think some systems have expense per-watcher like a background thread and kernel buffers - and those can approach a limit - so more of them uses more resources or could hit the limit - thus the whole polling watcher.  From the same token those kernel buffers associated with a watcher limit how many events it can buffer between callbacks.  At least for servicing - let's try to keep the same number of watchers as before.  Folks can always set the provider if they want to experiment with a single FSW for perf.  If you can see a better option for vNext we can consider that.

Agreed that RefCounting is too big and complex.  I think a clear assignment of responsibility of lifetime for the file-provider is what we want.  If it's possible to transfer the responsibility to the source *and* have the source's lifetime managed by something we could do that in the future.